animation_nodes/algorithms/mesh/triangulate_mesh.pyx

```diff
@@ -0,0 +1,316 @@
+    if sign < 0.0: return -(angle - angle % 0.001)
+    return (angle - angle % 0.001)
+
+# Find the index of minimum angle.
+cdef int findEarHasMinAngle(Py_ssize_t polyLength, Vertex* verticesData):
+    cdef float angleMin = 3.14
+    cdef float angle
+    cdef Py_ssize_t i, angleMinIndex
+
+    angleMinIndex = 0
+    for i in range(polyLength):
+        if verticesData[i].isEar:
+            angle = abs(verticesData[i].angle)
+            if angleMin > angle:
+                angleMin = angle
+                angleMinIndex = i
+
+    return angleMinIndex
+
+# Checking points (reflex type) lies inside the new triangle.
+cdef bint pointsInTriangle(Py_ssize_t polyLength, Vertex* verticesData, Py_ssize_t earIndex):
+    cdef Py_ssize_t previousIndex = verticesData[earIndex].previous
+    cdef Py_ssize_t nextIndex = verticesData[earIndex].next
+
+    cdef Vector3 v1 = verticesData[previousIndex].location
+    cdef Vector3 v2 = verticesData[earIndex].location
+    cdef Vector3 v3 = verticesData[nextIndex].location
+
+    cdef Py_ssize_t i
+    for i in range(polyLength):
+        if i != previousIndex and i != earIndex and i != nextIndex and pointInsideTriangle(v1, v2, v3, verticesData[i].location):
+            return True
+    return False
+
+@cython.cdivision(True)
+cdef bint pointInsideTriangle(Vector3 v1, Vector3 v2, Vector3 v3, Vector3 p):
+    cdef Vector3 u, v, w, n
+    subVec3(&u, &v2, &v1)
+    subVec3(&v, &v3, &v1)
+    subVec3(&w, &p, &v1)
+
+    cdef float delta = 0.00001
+    v.x += delta
+    v.y += delta
+    v.z += delta
+
+    crossVec3(&n, &u, &v)
+    cdef float normSq = lengthVec3(&n)
+    normSq = normSq * normSq
+
+    cdef Vector3 uw, wv
+    crossVec3(&uw, &u, &w)
+    crossVec3(&wv, &w, &v)
+
+    cdef float alpha = dotVec3(&uw, &n) / normSq
+    cdef float beta = dotVec3(&wv, &n) / normSq
+    cdef float gamma = 1.0 - alpha - beta
+    if alpha >= 0.0 and alpha <= 1.0 and beta >= 0.0 and beta <= 1.0 and gamma >= 0.0 and gamma <= 1.0:
+        return True
+    else:
+        return False
+
+# Transformation of vertices of polygon into xy-plane.
+@cython.cdivision(True)
+cdef void projectPolygonVertices(Vector3DList vertices, Py_ssize_t polyStart, Py_ssize_t polyLength,
+                                 unsigned int* oldIndices, Vertex* verticesData):
+    # Compute polygon normal with Nowell's method.
+    cdef Vector3 polyNormal = toVector3((0, 0, 0))
+    cdef Py_ssize_t i
+    for i in range(polyLength):
+        v1 = vertices.data[oldIndices[polyStart + i]]
+        v2 = vertices.data[oldIndices[polyStart + (i + 1) % polyLength]]
+
+        polyNormal.x += (v1.y - v2.y) * (v1.z + v2.z)
+        polyNormal.y += (v1.z - v2.z) * (v1.x + v2.x)
+        polyNormal.z += (v1.x - v2.x) * (v1.y + v2.y)
+
+        verticesData[i].index = oldIndices[polyStart + i]
+        verticesData[i].location = v1
+        verticesData[i].previous = (polyLength + i - 1) % polyLength
+        verticesData[i].next = (i + 1) % polyLength
```

Comment:
I prefer to handle the first and last vertex independently, to avoid the remainder operator here.

Reply:
It is done.